Spotted a weird test failure, not sure if it's related to the changes here or some latent flakiness:

```
2026-02-16T11:22:27.9025052Z [ERROR] Tests run: 12, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 9.484 s <<< FAILURE! -- in io.netty.handler.codec.http2.Http2MultiplexTransportTest
2026-02-16T11:22:27.9026228Z [ERROR] io.netty.handler.codec.http2.Http2MultiplexTransportTest -- Time elapsed: 9.484 s <<< ERROR!
2026-02-16T11:22:27.9027688Z java.lang.IllegalStateException: Possible memory leak detected for resource scope 'Http2MultiplexTransportTest'. Please use paranoid leak detection to get more information, or set -Dio.netty.util.LeakPresenceDetector.trackCreationStack=true
2026-02-16T11:22:27.9029306Z 	at io.netty.util.LeakPresenceDetector$ResourceScope.check(LeakPresenceDetector.java:307)
2026-02-16T11:22:27.9030210Z 	at io.netty.util.LeakPresenceDetector$ResourceScope.close(LeakPresenceDetector.java:327)
2026-02-16T11:22:27.9031607Z 	at io.netty.util.test.LeakPresenceExtension.afterAll(LeakPresenceExtension.java:108)
2026-02-16T11:22:27.9032636Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
2026-02-16T11:22:27.9033485Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
2026-02-16T11:22:27.9034347Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
2026-02-16T11:22:27.9035255Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
2026-02-16T11:22:27.9036233Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
2026-02-16T11:22:27.9036896Z 
2026-02-16T11:22:27.9314158Z [INFO] 
2026-02-16T11:22:27.9314516Z [INFO] Results:
2026-02-16T11:22:27.9314836Z [INFO] 
2026-02-16T11:22:27.9315348Z [ERROR] Errors: 
2026-02-16T11:22:27.9321828Z [ERROR]   Http2MultiplexTransportTest ? IllegalState Possible memory leak detected for resource scope 'Http2MultiplexTransportTest'. Please use paranoid leak detection to get more information, or set -Dio.netty.util.LeakPresenceDetector.trackCreationStack=true
```

--- a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
@@ -1374,6 +1374,14 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
         }
     }
 
+    private void decodeNonJdkCompatible(ChannelHandlerContext ctx, ByteBuf in) {
+        try {
+            unwrap(ctx, in, in.readableBytes());
+        } catch (Throwable cause) {
+            handleUnwrapThrowable(ctx, cause);
+        }
+    }
+
     private void handleUnwrapThrowable(ChannelHandlerContext ctx, Throwable cause) {
         try {
             // We should attempt to notify the handshake failure before writing any pending data. If we are in unwrap
@@ -1408,11 +1416,7 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
         if (jdkCompatibilityMode) {
             decodeJdkCompatible(ctx, in);
         } else {
-            try {
-                unwrap(ctx, in, in.readableBytes());
-            } catch (Throwable cause) {
-                handleUnwrapThrowable(ctx, cause);
-            }
+            decodeNonJdkCompatible(ctx, in);
         }
     }
 

--- a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
@@ -301,27 +301,26 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
             SSLEngineResult unwrap(SslHandler handler, ByteBuf in, int len, ByteBuf out) throws SSLException {
                 int writerIndex = out.writerIndex();
                 ByteBuffer inNioBuffer = getUnwrapInputBuffer(handler, toByteBuffer(in, in.readerIndex(), len));
+                int position = inNioBuffer.position();
                 final SSLEngineResult result = handler.engine.unwrap(inNioBuffer,
-                    toByteBuffer(out, writerIndex, out.writableBytes()));
+                        toByteBuffer(out, writerIndex, out.writableBytes()));
                 out.writerIndex(writerIndex + result.bytesProduced());
-                return result;
-            }
 
-            private ByteBuffer getUnwrapInputBuffer(SslHandler handler, ByteBuffer inNioBuffer) {
-                int javaVersion = PlatformDependent.javaVersion();
-                if (javaVersion >= 22 && javaVersion < 25 && inNioBuffer.isDirect()) {
-                    // Work-around for https://bugs.openjdk.org/browse/JDK-8357268
-                    int remaining = inNioBuffer.remaining();
-                    ByteBuffer copy = handler.unwrapInputCopy;
-                    if (copy == null || copy.capacity() < remaining) {
-                        handler.unwrapInputCopy = copy = ByteBuffer.allocate(remaining);
-                    } else {
-                        copy.clear();
+                // This is a workaround for a bug in Android 5.0. Android 5.0 does not correctly update the
+                // SSLEngineResult.bytesConsumed() in some cases and just return 0.
+                //
+                // See:
+                //     - https://android-review.googlesource.com/c/platform/external/conscrypt/+/122080
+                //     - https://github.com/netty/netty/issues/7758
+                if (result.bytesConsumed() == 0) {
+                    int consumed = inNioBuffer.position() - position;
+                    if (consumed != result.bytesConsumed()) {
+                        // Create a new SSLEngineResult with the correct bytesConsumed().
+                        return new SSLEngineResult(
+                                result.getStatus(), result.getHandshakeStatus(), consumed, result.bytesProduced());
                     }
-                    copy.put(inNioBuffer).flip();
-                    return copy;
                 }
-                return inNioBuffer;
+                return result;
             }
 
             @Override

--- a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
@@ -303,7 +303,7 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
                 ByteBuffer inNioBuffer = getUnwrapInputBuffer(handler, toByteBuffer(in, in.readerIndex(), len));
                 int position = inNioBuffer.position();
                 final SSLEngineResult result = handler.engine.unwrap(inNioBuffer,
-                        toByteBuffer(out, writerIndex, out.writableBytes()));
+                    toByteBuffer(out, writerIndex, out.writableBytes()));
                 out.writerIndex(writerIndex + result.bytesProduced());
 
                 // This is a workaround for a bug in Android 5.0. Android 5.0 does not correctly update the
@@ -321,6 +321,23 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
                     }
                 }
                 return result;
+            }
+
+            private ByteBuffer getUnwrapInputBuffer(SslHandler handler, ByteBuffer inNioBuffer) {
+                int javaVersion = PlatformDependent.javaVersion();
+                if (javaVersion >= 22 && javaVersion < 25 && inNioBuffer.isDirect()) {
+                    // Work-around for https://bugs.openjdk.org/browse/JDK-8357268
+                    int remaining = inNioBuffer.remaining();
+                    ByteBuffer copy = handler.unwrapInputCopy;
+                    if (copy == null || copy.capacity() < remaining) {
+                        handler.unwrapInputCopy = copy = ByteBuffer.allocate(remaining);
+                    } else {
+                        copy.clear();
+                    }
+                    copy.put(inNioBuffer).flip();
+                    return copy;
+                }
+                return inNioBuffer;
             }
 
             @Override

